src/main/java/com/datadoghq/sketch/ddsketch/encoding/ByteArrayInput.java

```diff
@@ -0,0 +1,63 @@
+/* Unless explicitly stated otherwise all files in this repository are licensed under the Apache License 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2021 Datadog, Inc.
+ */
+
+package com.datadoghq.sketch.ddsketch.encoding;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.Objects;
+
+/** An implementation of {@link Input} that is backed by an array. */
+public final class ByteArrayInput implements Input {
+
+  private final byte[] array;
+  private final int endPos;
+  private int pos;
+
+  public ByteArrayInput(byte[] array, int offset, int length) {
+    Objects.requireNonNull(array);
+    if (offset < 0 || offset + length > array.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    this.array = array;
+    this.endPos = offset + length;
+    this.pos = offset;
+  }
+
+  public ByteArrayInput(byte[] array) {
+    this(array, 0, array.length);
+  }
+
+  @Override
+  public final boolean hasRemaining() {
+    return pos < endPos;
+  }
+
+  @Override
+  public final byte readByte() throws EOFException {
+    if (pos >= endPos) {
+      throw new EOFException();
+    }
+    return array[pos++];
+  }
+
+  @Override
+  public final long readLongLE() throws IOException {
+    if (pos > endPos - 8) {
+      throw new EOFException();
+    }
+    long value = 0;
+    value |= Byte.toUnsignedLong(array[pos]);
+    value |= Byte.toUnsignedLong(array[pos + 1]) << 8;
+    value |= Byte.toUnsignedLong(array[pos + 2]) << 16;
+    value |= Byte.toUnsignedLong(array[pos + 3]) << 24;
+    value |= Byte.toUnsignedLong(array[pos + 4]) << 32;
+    value |= Byte.toUnsignedLong(array[pos + 5]) << 40;
+    value |= Byte.toUnsignedLong(array[pos + 6]) << 48;
+    value |= Byte.toUnsignedLong(array[pos + 7]) << 56;
+    pos += 8;
+    return value;
```

Comment:
While these get coalesced into a single load in go this is expensive on the JVM. Consider going multi-release for this - use `Unsafe` on JDK8 and `MethodHandles.byteArrayViewVarHandle` - [javadoc](https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandles.html#byteArrayViewVarHandle-java.lang.Class-java.nio.ByteOrder-).

Reply:
Another note - if you use `ByteBuffer.putLong` you get the right thing automatically

Reply:
>Another note - if you use ByteBuffer.putLong you get the right thing automatically

I may be mistaken, but looking at the implementation, I don't think Java 8 does it. Later versions do it however.

I'd like to keep those optimizations outside of this PR (so as not to block it), but [here](https://github.com/DataDog/sketches-java/pull/59) something I think we could do based on your feedback. Let me know if that's in line with what you have in mind. I'm also wondering about possible drawbacks of actually using `Unsafe` (I can see a couple of warnings when building).

Reply:
Yes, it was fixed since JDK8, precisely because assembling this operation can be slow.